### PR TITLE
Notify item changed in header and footer

### PR DESCRIPTION
**Pull request by yccheok**
Notify item changed in header and footer.

### CODE REVIEW

File: library/src/main/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapter.java [hunk 1]
```diff
@@ -451,6 +451,58 @@ public int getPositionInAdapter(Section section, int position) {
...
+
+    /**
+     * Helper method that returns the position of header in the adapter.
+     *
+     * @param section a visible section of this adapter
+     * @return position of the header in the adapter. -1 if the section doesn't have header
+     */
+    public int getHeaderPositionInAdapter(Section section) {
+        if (!section.hasHeader) {
+            return -1;
```
Comment (luizgrp) [inline library/src/main/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapter.java]: Please throw an IllegalStateException with message "Section doesn't have a header" and update javadoc accordingly

File: library/src/main/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapter.java [hunk 2]
```diff
@@ -451,6 +451,58 @@ public int getPositionInAdapter(Section section, int position) {
...
+
+    /**
+     * Helper method that returns the position of header in the adapter.
+     *
+     * @param section a visible section of this adapter
+     * @return position of the footer in the adapter. -1 if the section doesn't have footer
+     */
+    public int getFooterPositionInAdapter(Section section) {
+        if (!section.hasFooter) {
+            return -1;
```
Comment (luizgrp) [inline library/src/main/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapter.java]: Please throw an IllegalStateException with message "Section doesn't have a footer" and update javadoc accordingly.
Please also add a unit test to check if this method throws the exception.

File: library/src/main/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapter.java [hunk 3]
```diff
@@ -584,6 +636,70 @@ public void notifyItemChangedInSection(Section section, int position) {
...
     }
 
+    /**
+     * Helper method that calculates the relative header position in the adapter and calls
+     * {@link #notifyItemChanged}. Do nothing if the section doesn't have header
+     *
+     * @param tag unique identifier of the section
+     */
+    public void notifyHeaderChangedInSection(String tag) {
+        int headerPosition = getHeaderPositionInAdapter(tag);
```
Comment (luizgrp) [inline library/src/main/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapter.java]: any reason for this not calling `getValidSectionOrThrowException(tag)` then the overloaded method `notifyHeaderChangedInSection(Section)`?

File: library/src/main/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapter.java [hunk 4]
```diff
@@ -584,6 +636,70 @@ public void notifyItemChangedInSection(Section section, int position) {
...
+    /**
+     * Helper method that calculates the relative header position in the adapter and calls
+     * {@link #notifyItemChanged}. Do nothing if the section doesn't have header
+     *
+     * @param section a visible section of this adapter
+     */
+    public void notifyHeaderChangedInSection(Section section) {
+        int headerPosition = getHeaderPositionInAdapter(section);
+
+        if (headerPosition == -1) {
```
Comment (luizgrp) [inline library/src/main/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapter.java]: this check can be removed after changing `getHeaderPositionInAdapter` to throw exception

File: library/src/main/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapter.java [hunk 5]
```diff
@@ -584,6 +636,70 @@ public void notifyItemChangedInSection(Section section, int position) {
...
+    }
+
+    /**
+     * Helper method that calculates the relative header position in the adapter and calls
+     * {@link #notifyItemChanged}. Do nothing if the section doesn't have footer
+     *
+     * @param tag unique identifier of the section
+     */
+    public void notifyFooterChangedInSection(String tag) {
+        int footerPosition = getFooterPositionInAdapter(tag);
```
Comment (luizgrp) [inline library/src/main/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapter.java]: same comment as for `notifyHeaderChangedInSection(tag)`

File: library/src/main/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapter.java [hunk 6]
```diff
@@ -584,6 +636,70 @@ public void notifyItemChangedInSection(Section section, int position) {
...
+    /**
+     * Helper method that calculates the relative header position in the adapter and calls
+     * {@link #notifyItemChanged}. Do nothing if the section doesn't have header
+     *
+     * @param section a visible section of this adapter
+     */
+    public void notifyFooterChangedInSection(Section section) {
+        int footerPosition = getFooterPositionInAdapter(section);
+
+        if (footerPosition == -1) {
```
Comment (luizgrp) [inline library/src/main/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapter.java]: same comment as for `notifyHeaderChangedInSection(Section)`

Comment (yccheok) [inline library/src/main/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapter.java]: @luizgrp 

Thanks for review. Already fixed.

File: library/src/test/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapterTest.java [hunk 7]
```diff
@@ -460,6 +460,127 @@ public void getPositionInAdapterUsingSection_withAdapterWithManySections_returns
     }
 
     @Test
+    public void getHeaderPositionInAdapterUsingTag_withAdapterWithManySections_returnsCorrectAdapterHeaderPosition() {
+        // Given
+        SectionedRecyclerViewAdapter spySectionedRecyclerViewAdapter = spy(SectionedRecyclerViewAdapter.class);
+        doNothing().when(spySectionedRecyclerViewAdapter).callSuperNotifyItemInserted(anyInt());
```
Comment (luizgrp) [inline library/src/test/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapterTest.java]: there is no need to use a spy here, please use `sectionAdapter`

File: library/src/test/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapterTest.java [hunk 8]
```diff
@@ -460,6 +460,127 @@ public void getPositionInAdapterUsingSection_withAdapterWithManySections_returns
...
+
+        // Then
+        assertThat(result, is(10));
+    }
+
+    @Test(expected = IllegalStateException.class)
+    public void getHeaderPositionInAdapterUsingTag_withAdapterWithManySections_throwsIllegalStateException() {
+        // Given
+        SectionedRecyclerViewAdapter spySectionedRecyclerViewAdapter = spy(SectionedRecyclerViewAdapter.class);
+        doNothing().when(spySectionedRecyclerViewAdapter).callSuperNotifyItemInserted(anyInt());
```
Comment (luizgrp) [inline library/src/test/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapterTest.java]: same comment as above regarding spy

File: library/src/test/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapterTest.java [hunk 9]
```diff
@@ -460,6 +460,127 @@ public void getPositionInAdapterUsingSection_withAdapterWithManySections_returns
...
+
+        spySectionedRecyclerViewAdapter.addSection(new StatelessSectionStub(ITEMS_QTY));
+        spySectionedRecyclerViewAdapter.addSection(SECTION_TAG, new StatelessSectionStub(ITEMS_QTY));
+
+        // When
+        int result = spySectionedRecyclerViewAdapter.getHeaderPositionInAdapter(SECTION_TAG);
+    }
+
+    @Test
+    public void getHeaderPositionInAdapterUsingSection_withAdapterWithManySections_returnsCorrectAdapterHeaderPosition() {
```
Comment (luizgrp) [inline library/src/test/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapterTest.java]: Redundant test, everything tested in here is being tested already `getHeaderPositionInAdapterUsingTag_withAdapterWithManySections_returnsCorrectAdapterHeaderPosition`, please remove it.

File: library/src/test/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapterTest.java [hunk 10]
```diff
@@ -460,6 +460,127 @@ public void getPositionInAdapterUsingSection_withAdapterWithManySections_returns
...
+
+        // When
+        int result = spySectionedRecyclerViewAdapter.getHeaderPositionInAdapter(statelessSectionStub);
+    }
+
+    @Test
+    public void getFooterPositionInAdapterUsingTag_withAdapterWithManySections_returnsCorrectAdapterFooterPosition() {
+        // Given
+        SectionedRecyclerViewAdapter spySectionedRecyclerViewAdapter = spy(SectionedRecyclerViewAdapter.class);
+        doNothing().when(spySectionedRecyclerViewAdapter).callSuperNotifyItemInserted(anyInt());
```
Comment (luizgrp) [inline library/src/test/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapterTest.java]: same comment as above regarding spy

File: library/src/test/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapterTest.java [hunk 11]
```diff
@@ -460,6 +460,127 @@ public void getPositionInAdapterUsingSection_withAdapterWithManySections_returns
...
+
+        // Then
+        assertThat(result, is(21));
+    }
+
+    @Test(expected = IllegalStateException.class)
+    public void getFooterPositionInAdapterUsingTag_withAdapterWithManySections_throwsIllegalStateException() {
+        // Given
+        SectionedRecyclerViewAdapter spySectionedRecyclerViewAdapter = spy(SectionedRecyclerViewAdapter.class);
+        doNothing().when(spySectionedRecyclerViewAdapter).callSuperNotifyItemInserted(anyInt());
```
Comment (luizgrp) [inline library/src/test/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapterTest.java]: same comment as above regarding spy

File: library/src/test/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapterTest.java [hunk 12]
```diff
@@ -460,6 +460,127 @@ public void getPositionInAdapterUsingSection_withAdapterWithManySections_returns
...
+
+        spySectionedRecyclerViewAdapter.addSection(new StatelessSectionStub(ITEMS_QTY));
+        spySectionedRecyclerViewAdapter.addSection(SECTION_TAG, new StatelessSectionStub(ITEMS_QTY));
+
+        // When
+        int result = spySectionedRecyclerViewAdapter.getFooterPositionInAdapter(SECTION_TAG);
+    }
+
+    @Test
+    public void getFooterPositionInAdapterUsingSection_withAdapterWithManySections_returnsCorrectAdapterFooterPosition() {
```
Comment (luizgrp) [inline library/src/test/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapterTest.java]: redundant test

File: library/src/test/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapterTest.java [hunk 13]
```diff
@@ -625,6 +746,124 @@ public void notifyItemChangedInSectionUsingSection_withAdapterWithManySections_c
...
+
+        // When
+        spySectionedRecyclerViewAdapter.notifyHeaderChangedInSection(SECTION_TAG);
+
+        // Then
+        verify(spySectionedRecyclerViewAdapter).callSuperNotifyItemChanged(10);
+    }
+
+    @Test
+    public void notifyHeaderChangedInSectionUsingSection_withAdapterWithManySections_callsSuperNotifyItemChangedInSection() {
```
Comment (luizgrp) [inline library/src/test/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapterTest.java]: redundant test

File: library/src/test/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapterTest.java [hunk 14]
```diff
@@ -625,6 +746,124 @@ public void notifyItemChangedInSectionUsingSection_withAdapterWithManySections_c
...
+
+        // Then
+        verify(spySectionedRecyclerViewAdapter).callSuperNotifyItemChanged(10);
+    }
+
+    @Test(expected = IllegalStateException.class)
+    public void notifyHeaderChangedInSectionUsingTag_withAdapterWithManySections_throwsIllegalStateException() {
+        // Given
+        SectionedRecyclerViewAdapter spySectionedRecyclerViewAdapter = spy(SectionedRecyclerViewAdapter.class);
+        doNothing().when(spySectionedRecyclerViewAdapter).callSuperNotifyItemChanged(anyInt());
```
Comment (luizgrp) [inline library/src/test/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapterTest.java]: no need for a spy

File: library/src/test/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapterTest.java [hunk 15]
```diff
@@ -625,6 +746,124 @@ public void notifyItemChangedInSectionUsingSection_withAdapterWithManySections_c
...
+
+        spySectionedRecyclerViewAdapter.addSection(new StatelessSectionStub(ITEMS_QTY));
+        spySectionedRecyclerViewAdapter.addSection(SECTION_TAG, new StatelessSectionStub(ITEMS_QTY));
+
+        // When
+        spySectionedRecyclerViewAdapter.notifyHeaderChangedInSection(SECTION_TAG);
+    }
+
+    @Test(expected = IllegalStateException.class)
+    public void notifyHeaderChangedInSectionUsingSection_withAdapterWithManySections_throwsIllegalStateException() {
```
Comment (luizgrp) [inline library/src/test/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapterTest.java]: redundant test

File: library/src/test/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapterTest.java [hunk 16]
```diff
@@ -625,6 +746,124 @@ public void notifyItemChangedInSectionUsingSection_withAdapterWithManySections_c
...
+
+        // When
+        spySectionedRecyclerViewAdapter.notifyFooterChangedInSection(SECTION_TAG);
+
+        // Then
+        verify(spySectionedRecyclerViewAdapter).callSuperNotifyItemChanged(21);
+    }
+
+    @Test
+    public void notifyFooterChangedInSectionUsingSection_withAdapterWithManySections_callsSuperNotifyItemChangedInSection() {
```
Comment (luizgrp) [inline library/src/test/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapterTest.java]: redundant test

File: library/src/test/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapterTest.java [hunk 17]
```diff
@@ -625,6 +746,124 @@ public void notifyItemChangedInSectionUsingSection_withAdapterWithManySections_c
...
+
+        // Then
+        verify(spySectionedRecyclerViewAdapter).callSuperNotifyItemChanged(21);
+    }
+
+    @Test(expected = IllegalStateException.class)
+    public void notifyFooterChangedInSectionUsingTag_withAdapterWithManySections_throwsIllegalStateException() {
+        // Given
+        SectionedRecyclerViewAdapter spySectionedRecyclerViewAdapter = spy(SectionedRecyclerViewAdapter.class);
+        doNothing().when(spySectionedRecyclerViewAdapter).callSuperNotifyItemChanged(anyInt());
```
Comment (luizgrp) [inline library/src/test/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapterTest.java]: no need for a spy

File: library/src/test/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapterTest.java [hunk 18]
```diff
@@ -625,6 +746,124 @@ public void notifyItemChangedInSectionUsingSection_withAdapterWithManySections_c
...
+
+        spySectionedRecyclerViewAdapter.addSection(new StatelessSectionStub(ITEMS_QTY));
+        spySectionedRecyclerViewAdapter.addSection(SECTION_TAG, new StatelessSectionStub(ITEMS_QTY));
+
+        // When
+        spySectionedRecyclerViewAdapter.notifyFooterChangedInSection(SECTION_TAG);
+    }
+
+    @Test(expected = IllegalStateException.class)
+    public void notifyFooterChangedInSectionUsingSection_withAdapterWithManySections_throwsIllegalStateException() {
```
Comment (luizgrp) [inline library/src/test/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapterTest.java]: redundant test

**Comment by yccheok** [comment]
@luizgrp Thanks. I hope I didn't accidentally deleted something after the fix

File: library/src/main/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapter.java [hunk 19]
```diff
@@ -584,6 +636,54 @@ public void notifyItemChangedInSection(Section section, int position) {
         callSuperNotifyItemChanged(getPositionInAdapter(section, position));
     }
 
+    /**
+     * Helper method that calculates the relative header position in the adapter and calls
+     * {@link #notifyItemChanged}. Do nothing if the section doesn't have header
```
Comment (luizgrp) [inline library/src/main/java/io/github/luizgrp/sectionedrecyclerviewadapter/SectionedRecyclerViewAdapter.java]: "do nothing" comments here and below are outdated, I will remove them in develop after merge

**Pull request merged**